src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -190,42 +206,76 @@ private ReplicationBatch[] PrepareReplicationDataForShards(TransactionOperationC
             {
                 foreach (var attachment in dataForReplicationCommand.ReplicatedAttachmentStreams)
                 {
+                    DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Make '_tempFileCache.MemoryStreamCapacity' configurable");
+                    var stream = attachment.Value.Size > _tempFileCache.MemoryStreamCapacity
+                            ? _tempFileCache.RentFileStream()
+                            : _tempFileCache.RentMemoryStream();
+
+                    _batches.Streams.Add(attachment.Key, stream);
+
+                    attachment.Value.Stream.Seek(0, SeekOrigin.Begin);
+                    attachment.Value.Stream.CopyTo(stream);
+
                     for (var shard = 0; shard < _parent.Context.ShardCount; shard++)
                     {
                         if (batches[shard].Attachments?.ContainsKey(attachment.Key) != true)
                             continue;
 
-                        var buffer = GetBufferFromAttachmentStream(attachment.Value.Stream);
-
-                        var attachmentStream = new AttachmentReplicationItem
+                        AddAttachmentStreamItemToBatch(batches[shard], attachment.Key, stream);
+                    }
+                }
+            }
+            else
+            {
+                foreach (var batch in batches)
+                {
+                    if (batch.Attachments != null)
+                    {
+                        foreach (var attachment in batch.Attachments)
                         {
-                            Type = ReplicationBatchItem.ReplicationItemType.AttachmentStream,
-                            Base64Hash = attachment.Key,
-                            Stream = new MemoryStream(buffer)
-                        };
-
-                        batches[shard].Attachments[attachment.Key] = attachmentStream;
-
-                        DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Optimization required");
+                            if (attachment.Value == null && _batches.Streams.ContainsKey(attachment.Key))
```

Comment:
in this code path, when `_batches.Streams` can contain anything? you are getting here when `dataForReplicationCommand.ReplicatedAttachmentStreams` is `null`

Reply:
Removed